feat(release): Deploy built spinnaker version to k8s.

jtk54 opened the pull request:
This installs Halyard on an ubuntu:14.04 container and subsequently deploys Spinnaker on the same k8s cluster using Halyard. Once Spinnaker is running on k8s, we can tunnel using `kubectl` to run a citest suite.

The container is built using something like:
```
docker build --build-arg BUILD_PROJECT="${project}" \
   --build-arg  BUILD_ACCOUNT="${builder account}" \
   --build-arg BUILD_KEY="$(cat build.json)" \
   --build-arg DOCKER_PROJECT="${docker project}" \
   --build-arg DOCKER_ACCOUNT="${docker project account}" \
   --build-arg DOCKER_KEY="$(cat docker.json)" \
   --build-arg CLUSTER_NAME="halyard-cluster" \
   --build-arg KUBE_CONF="$(python -c 'import sys, yaml, json; json.dump(yaml.load(sys.stdin), sys.stdout, indent=4)' < ~/.kube/config)" \
   -t ${tag name} -f hal_install_spin.docker .
```

and subsequently run with `kubectl create -f ${pod spec}.yml` after the building and pushing the container.

lwander commented on dev/hal_install_spin.docker:
```diff
@@ -0,0 +1,34 @@
+# Parameters of the GCP project Spinnaker is maintaining.
+ARG BUILD_PROJECT
+ARG BUILD_ACCOUNT
+ARG BUILD_KEY # json keyfile contents
```
Secrets shouldn't be build-time args, since that would have us baking keys into images.

lwander replied:
It's totally OK to assume (& document) what env vars need to be set in the container at runtime. You can pass these in either via the docker CLI, or by modifying the pod spec that's running this container in k8s.

lwander replied:
You can also load env vars in k8s from secrets, which would probably be the golden-path here. 

`  kubectl create secret generic kubeconfig --from-file=relative/path/to/kube/config`

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: hal
spec:
  containers:
    - name: hal
      image: hal
      env:
        - name: KUBE_CONFIG
          valueFrom:
            secretKeyRef:
              name: kubeconfig
              key: config # this is the name of the file you published in the secret
```

jtk54 replied:
Moved stuff to secrets.

lwander commented on dev/hal_k8s_run.sh:
```diff
@@ -0,0 +1,86 @@
+
+sudo bash InstallHalyard.sh -y
+echo "Configuring Halyard..."
+cat > /opt/spinnaker/config/halyard.yml  <<EOF
```
Are you publishing BOMs? if not, the default halyard.yml that comes with your installation should be enough

jtk54 replied:
Good point.

jtk54 commented:
I updated this to use secrets instead of injecting environment variables, and removed unnecessary environment variables. This container is now built with the usual `docker build` (no `--build-arg`s), but needs to be run with a pod spec similar to (via @lwander):

```
apiVersion: v1
kind: Pod
metadata:
  name: hal
spec:
  containers:
    - name: hal
      image: hal
      env:
        - name: KUBE_CONFIG
          valueFrom:
            secretKeyRef:
              name: kubeconfig
              key: config # this is the name of the file you published in the secret
```
with an `env` entry for each of the required environment variables.

lwander commented:
sweet

lwander commented on dev/hal_install_spin.docker:
```diff
@@ -0,0 +1,26 @@
+
+RUN apt-get install -y curl software-properties-common apt-transport-https wget
+
+# The container expects the following environment variables to be set.
```
looks good
